File: android/src/main/kotlin/dev/luix/connect_kit/mapper/CategoryMapper.kt
package dev.luix.connect_kit.mapper

import androidx.health.connect.client.records.BodyTemperatureMeasurementLocation
import androidx.health.connect.client.records.BloodGlucoseRecord
import androidx.health.connect.client.records.MealType
import androidx.health.connect.client.records.Vo2MaxRecord
import androidx.health.connect.client.records.SkinTemperatureRecord
import androidx.health.connect.client.records.CervicalMucusRecord
import androidx.health.connect.client.records.MindfulnessSessionRecord
import androidx.health.connect.client.records.MenstruationFlowRecord
import androidx.health.connect.client.records.OvulationTestRecord
import androidx.health.connect.client.records.SexualActivityRecord
import androidx.health.connect.client.records.ActivityIntensityRecord
import androidx.health.connect.client.records.SleepSessionRecord

/**
 * Maps category values between Dart enum strings and Android Health Connect constants.
 *
 * This mapper provides bidirectional conversions:
 *  - [decode] converts a Dart enum string (e.g. "wrist") into the Android Int constant.
 *  - [encode] converts an Android Int constant back into a Dart enum string.
 *
 * Internally, each category defines its own map of `String` ↔ `Int`.
 */
object CategoryMapper {

    // ------------------------------------------------------------------------
    // Category mappings
    // ------------------------------------------------------------------------

    private val bodyTemperatureMeasurementLocation: Map<String, Int> = mapOf(
        "armpit" to BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_ARMPIT,
        "ear" to BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_EAR,
        "finger" to BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_FINGER,
        "forehead" to BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_FOREHEAD,
        "mouth" to BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_MOUTH,
        "rectum" to BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_RECTUM,
        "artery" to BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_TEMPORAL_ARTERY,
        "toe" to BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_TOE,
        "unknown" to BodyTemperatureMeasurementLocation.MEASUREMENT_LOCATION_UNKNOWN,
    )

    private val specimenSource: Map<String, Int> = mapOf(
        "capillaryBlood" to BloodGlucoseRecord.SPECIMEN_SOURCE_CAPILLARY_BLOOD,
        "interstitialFluid" to BloodGlucoseRecord.SPECIMEN_SOURCE_INTERSTITIAL_FLUID,
        "plasma" to BloodGlucoseRecord.SPECIMEN_SOURCE_PLASMA,
        "serum" to BloodGlucoseRecord.SPECIMEN_SOURCE_SERUM,
        "tears" to BloodGlucoseRecord.SPECIMEN_SOURCE_TEARS,
        "blood" to BloodGlucoseRecord.SPECIMEN_SOURCE_WHOLE_BLOOD,
        "unknown" to BloodGlucoseRecord.SPECIMEN_SOURCE_UNKNOWN,
    )

    private val mealType: Map<String, Int> = mapOf(
        "breakfast" to MealType.MEAL_TYPE_BREAKFAST,
        "dinner" to MealType.MEAL_TYPE_DINNER,
        "lunch" to MealType.MEAL_TYPE_LUNCH,
        "snack" to MealType.MEAL_TYPE_SNACK,
        "unknown" to MealType.MEAL_TYPE_UNKNOWN,
    )

    private val relationToMeal: Map<String, Int> = mapOf(
        "afterMeal" to BloodGlucoseRecord.RELATION_TO_MEAL_AFTER_MEAL,
        "beforeMeal" to BloodGlucoseRecord.RELATION_TO_MEAL_BEFORE_MEAL,
        "fasting" to BloodGlucoseRecord.RELATION_TO_MEAL_FASTING,
        "general" to BloodGlucoseRecord.RELATION_TO_MEAL_GENERAL,
        "unknown" to BloodGlucoseRecord.RELATION_TO_MEAL_UNKNOWN,
    )

    private val vo2MaxMeasurementMethod: Map<String, Int> = mapOf(
        "cooperTest" to Vo2MaxRecord.MEASUREMENT_METHOD_COOPER_TEST,
        "rateRatio" to Vo2MaxRecord.MEASUREMENT_METHOD_HEART_RATE_RATIO,
        "metabolicCart" to Vo2MaxRecord.MEASUREMENT_METHOD_METABOLIC_CART,
        "multistageFitnessTest" to Vo2MaxRecord.MEASUREMENT_METHOD_MULTISTAGE_FITNESS_TEST,
        "fitnessTest" to Vo2MaxRecord.MEASUREMENT_METHOD_ROCKPORT_FITNESS_TEST,
        "other" to Vo2MaxRecord.MEASUREMENT_METHOD_OTHER,
    )

    private val skinTemperatureMeasurementLocation: Map<String, Int> = mapOf(
        "finger" to SkinTemperatureRecord.MEASUREMENT_LOCATION_FINGER,
        "toe" to SkinTemperatureRecord.MEASUREMENT_LOCATION_TOE,
        "wrist" to SkinTemperatureRecord.MEASUREMENT_LOCATION_WRIST,
        "unknown" to SkinTemperatureRecord.MEASUREMENT_LOCATION_UNKNOWN,
    )

    private val mindfulnessSessionType: Map<String, Int> = mapOf(
        "breathing" to MindfulnessSessionRecord.MINDFULNESS_SESSION_TYPE_BREATHING,
        "meditation" to MindfulnessSessionRecord.MINDFULNESS_SESSION_TYPE_MEDITATION,
        "movement" to MindfulnessSessionRecord.MINDFULNESS_SESSION_TYPE_MOVEMENT,
        "music" to MindfulnessSessionRecord.MINDFULNESS_SESSION_TYPE_MUSIC,
        "unguided" to MindfulnessSessionRecord.MINDFULNESS_SESSION_TYPE_UNGUIDED,
        "unknown" to MindfulnessSessionRecord.MINDFULNESS_SESSION_TYPE_UNKNOWN,
    )

    private val menstruationFlow: Map<String, Int> = mapOf(
        "heavy" to MenstruationFlowRecord.FLOW_HEAVY,
        "light" to MenstruationFlowRecord.FLOW_LIGHT,
        "medium" to MenstruationFlowRecord.FLOW_MEDIUM,
        "unknown" to MenstruationFlowRecord.FLOW_UNKNOWN,
    )

    private val cervicalMucusAppearance: Map<String, Int> = mapOf(
        "dry" to CervicalMucusRecord.APPEARANCE_DRY,
        "sticky" to CervicalMucusRecord.APPEARANCE_STICKY,
        "creamy" to CervicalMucusRecord.APPEARANCE_CREAMY,
        "watery" to CervicalMucusRecord.APPEARANCE_WATERY,
        "eggWhite" to CervicalMucusRecord.APPEARANCE_EGG_WHITE,
        "unknown" to CervicalMucusRecord.APPEARANCE_UNKNOWN,
    )

    private val cervicalMucusSensation: Map<String, Int> = mapOf(
        "light" to CervicalMucusRecord.SENSATION_LIGHT,
        "medium" to CervicalMucusRecord.SENSATION_MEDIUM,
        "heavy" to CervicalMucusRecord.SENSATION_HEAVY,
        "unknown" to CervicalMucusRecord.SENSATION_UNKNOWN,
    )

    private val ovulationTestResult: Map<String, Int> = mapOf(
        "high" to OvulationTestRecord.RESULT_HIGH,
        "negative" to OvulationTestRecord.RESULT_NEGATIVE,
        "positive" to OvulationTestRecord.RESULT_POSITIVE,
        "inconclusive" to OvulationTestRecord.RESULT_INCONCLUSIVE,
    )

    private val sexualActivityProtection: Map<String, Int> = mapOf(
        "protected" to SexualActivityRecord.PROTECTION_USED_PROTECTED,
        "unprotected" to SexualActivityRecord.PROTECTION_USED_UNPROTECTED,
        "unknown" to SexualActivityRecord.PROTECTION_USED_UNKNOWN,
    )

    private val activityIntensityType: Map<String, Int> = mapOf(
        "moderate" to ActivityIntensityRecord.ACTIVITY_INTENSITY_TYPE_MODERATE,
        "vigorous" to ActivityIntensityRecord.ACTIVITY_INTENSITY_TYPE_VIGOROUS,
    )

    private val sleepSession: Map<String, Int> = mapOf(
        "inBed" to SleepSessionRecord.STAGE_TYPE_AWAKE_IN_BED,
        "outOfBed" to SleepSessionRecord.STAGE_TYPE_OUT_OF_BED,
        "sleeping" to SleepSessionRecord.STAGE_TYPE_SLEEPING,
        "awake" to SleepSessionRecord.STAGE_TYPE_AWAKE,
        "light" to SleepSessionRecord.STAGE_TYPE_LIGHT,
        "deep" to SleepSessionRecord.STAGE_TYPE_DEEP,
        "rem" to SleepSessionRecord.STAGE_TYPE_REM,
        "unknown" to SleepSessionRecord.STAGE_TYPE_UNKNOWN,
    )

    /**
     * Registry using Dart enum class names as keys (from runtimeType.toString()).
     *
     * Keys must match the Dart enum class names exactly (e.g., "CKMealType", "CKSpecimenSource").
     */
    private val registry: Map<String, Map<String, Int>> = mapOf(
        "CKBodyTemperatureMeasurementLocation" to bodyTemperatureMeasurementLocation,
        "CKSpecimenSource" to specimenSource,
        "CKMealType" to mealType,
        "CKRelationToMeal" to relationToMeal,
        "CKVo2MaxMeasurementMethod" to vo2MaxMeasurementMethod,
        "CKSkinTemperatureMeasurementLocation" to skinTemperatureMeasurementLocation,
        "CKMindfulnessSessionType" to mindfulnessSessionType,
        "CKMenstruationFlow" to menstruationFlow,
        "CKCervicalMucusAppearance" to cervicalMucusAppearance,
        "CKCervicalMucusSensation" to cervicalMucusSensation,
        "CKOvulationTestResult" to ovulationTestResult,
        "CKSexualActivityProtection" to sexualActivityProtection,
        "CKActivityIntensityType" to activityIntensityType,
        // NOTE: SleepSession is not a category enum in dart (no `CK` prefix)
        "SleepSession" to sleepSession, // Note: This might need to be CKSleepStage depending on Dart implementation
    )

    // ------------------------------------------------------------------------
    // Public API
    // ------------------------------------------------------------------------

    /**
     * Decodes a Dart enum string into its corresponding Android Int constant.
     *
     * @param category The category name (e.g. "SkinTemperatureMeasurementLocation").
     * @param value The Dart enum value string (e.g. "wrist").
     * @return The matching Android Int constant, or `null` if unknown.
     * @param categoryName The Dart enum class name (e.g. "CKMealType") - injected from Dart via runtimeType.toString()
     * @param value The Dart enum value string (e.g. "breakfast")
     * @return The matching Android Int constant, or `null` if unknown
     */
    fun decode(categoryName: String?, value: String?): Int? {
        if (categoryName == null || value == null) return null
        val categoryMap = registry[categoryName] ?: return null
        return categoryMap[value]
    }

    /**
     * Encodes an Android Int constant into its corresponding Dart enum string.
     *
     * @param categoryName The Dart enum class name (e.g. "CKMealType")
     * @param constant The Android constant value (e.g. MEAL_TYPE_BREAKFAST)
     * @return The Dart enum value string (e.g. "breakfast"), or `null` if unknown
     */
    fun encode(categoryName: String, constant: Int): String? {
        val categoryMap = registry[categoryName] ?: return null
        return categoryMap.entries.firstOrNull { it.value == constant }?.key
    }
}

File: ios/Classes/Mapper/CategoryMapper.swift
import Foundation
import HealthKit

/// Maps category names and values between Dart strings and iOS HealthKit constants.
///
/// This mapper provides bidirectional conversions:
///  - `decode` converts a Dart enum string (e.g. "wrist") into the iOS Int constant.
///  - `encode` converts an iOS Int constant back into a Dart enum string.
///
/// Internally, each category defines its own map of `String` ↔ `Int`.
/// Mirrors Android CategoryMapper.kt exactly for cross-platform consistency.
public enum CategoryMapper {

    // ------------------------------------------------------------------------
    // Category mappings
    // ------------------------------------------------------------------------

    private static let biologicalSex: [String: Int] = [
        "female": HKBiologicalSex.female.rawValue,
        "male": HKBiologicalSex.male.rawValue,
        "other": HKBiologicalSex.other.rawValue,  // @available(macOS 13.0, *)
        "unknown": HKBiologicalSex.notSet.rawValue,
    ]

    private static let bloodType: [String: Int] = [
        "aPositive": HKBloodType.aPositive.rawValue,
        "aNegative": HKBloodType.aNegative.rawValue,
        "bPositive": HKBloodType.bPositive.rawValue,
        "bNegative": HKBloodType.bNegative.rawValue,
        "abPositive": HKBloodType.abPositive.rawValue,
        "abNegative": HKBloodType.abNegative.rawValue,
        "oPositive": HKBloodType.oPositive.rawValue,
        "oNegative": HKBloodType.oNegative.rawValue,
        "unknown": HKBloodType.notSet.rawValue,
    ]

    private static let fitzpatrickSkinType: [String: Int] = [
        "I": HKFitzpatrickSkinType.I.rawValue,
        "II": HKFitzpatrickSkinType.II.rawValue,
        "III": HKFitzpatrickSkinType.III.rawValue,
        "IV": HKFitzpatrickSkinType.IV.rawValue,
        "V": HKFitzpatrickSkinType.V.rawValue,
        "VI": HKFitzpatrickSkinType.VI.rawValue,
        "unknown": HKFitzpatrickSkinType.notSet.rawValue,
    ]

    private static let intermenstrualBleedingType: [String: Int] = [
        "notApplicable": HKCategoryValue.notApplicable.rawValue,
        "unknown": HKCategoryValue.notApplicable.rawValue,
    ]

    private static let contraceptiveValue: [String: Int] = {
        // iOS 14.3+ — full map available
        if #available(iOS 14.3, *) {
            return [
                "unspecified": HKCategoryValueContraceptive.unspecified.rawValue,
                "implant": HKCategoryValueContraceptive.implant.rawValue,
                "injection": HKCategoryValueContraceptive.injection.rawValue,
                "intrauterineDevice": HKCategoryValueContraceptive.intrauterineDevice.rawValue,
                "intravaginalRing": HKCategoryValueContraceptive.intravaginalRing.rawValue,
                "oral": HKCategoryValueContraceptive.oral.rawValue,
                "patch": HKCategoryValueContraceptive.patch.rawValue,
                "unknown": HKCategoryValueContraceptive.unspecified.rawValue,
            ]
        }

        // Pre–iOS 14.3 — None of these symbols exist (it is guaranteed by the RecordTypeMapper)
        return [:]
    }()

    private static let progesteroneTestResult: [String: Int] = {
        // iOS 15.0+ — full map available
        if #available(iOS 15.0, *) {
            return [
                "negative": HKCategoryValueProgesteroneTestResult.negative.rawValue,
                "positive": HKCategoryValueProgesteroneTestResult.positive.rawValue,
                "indeterminate": HKCategoryValueProgesteroneTestResult.indeterminate.rawValue,
                "unknown": HKCategoryValueProgesteroneTestResult.indeterminate.rawValue,
            ]
        }

        // Pre–iOS 15.0 — None of these symbols exist (it is guaranteed by the RecordTypeMapper)
        return [:]
    }()

    /// Body temperature measurement location mappings
    private static let bodyTemperatureMeasurementLocation: [String: Int] = [
        "armpit": HKBodyTemperatureSensorLocation.armpit.rawValue,
        "body": HKBodyTemperatureSensorLocation.body.rawValue,
        "ear": HKBodyTemperatureSensorLocation.ear.rawValue,
        "finger": HKBodyTemperatureSensorLocation.finger.rawValue,
        "gastroIntestinal": HKBodyTemperatureSensorLocation.gastroIntestinal.rawValue,
        "mouth": HKBodyTemperatureSensorLocation.mouth.rawValue,
        "rectum": HKBodyTemperatureSensorLocation.rectum.rawValue,
        "toe": HKBodyTemperatureSensorLocation.toe.rawValue,
        "earDrum": HKBodyTemperatureSensorLocation.earDrum.rawValue,
        "artery": HKBodyTemperatureSensorLocation.temporalArtery.rawValue,
        "forehead": HKBodyTemperatureSensorLocation.forehead.rawValue,
        "unknown": HKBodyTemperatureSensorLocation.other.rawValue,
    ]

    private static let mindfulnessSessionType: [String: Int] = [
        "unknown": HKCategoryValue.notApplicable.rawValue
    ]

    /// Menstruation flow mappings
    private static let menstruationFlow: [String: Int] = {
        // iOS 18.0+ — full map available
        if #available(iOS 18.0, *) {
            return [
                "light": HKCategoryValueVaginalBleeding.light.rawValue,
                "medium": HKCategoryValueVaginalBleeding.medium.rawValue,
                "heavy": HKCategoryValueVaginalBleeding.heavy.rawValue,
                "none": HKCategoryValueVaginalBleeding.none.rawValue,
                "unknown": HKCategoryValueVaginalBleeding.unspecified.rawValue,
            ]
        }

        // Pre–iOS 18.0 — None of these symbols exist (it is guaranteed by the RecordTypeMapper)
        return [:]
    }()

    /// Cervical mucus appearance mappings
    private static let cervicalMucusAppearance: [String: Int] = [
        "dry": HKCategoryValueCervicalMucusQuality.dry.rawValue,
        "sticky": HKCategoryValueCervicalMucusQuality.sticky.rawValue,
        "creamy": HKCategoryValueCervicalMucusQuality.creamy.rawValue,
        "watery": HKCategoryValueCervicalMucusQuality.watery.rawValue,
        "eggWhite": HKCategoryValueCervicalMucusQuality.eggWhite.rawValue,
        "unknown": HKCategoryValueCervicalMucusQuality.dry.rawValue,  // no unspecified in iOS
    ]

    /// Ovulation test result mappings
    private static let ovulationTestResult: [String: Int] = [
        "high": HKCategoryValueOvulationTestResult.estrogenSurge.rawValue,
        "negative": HKCategoryValueOvulationTestResult.negative.rawValue,
        "positive": HKCategoryValueOvulationTestResult.luteinizingHormoneSurge.rawValue,
        "inconclusive": HKCategoryValueOvulationTestResult.indeterminate.rawValue,
    ]

    /// Sexual activity protection mappings
    /// Documentation says to store HKCategoryValue.notApplicable, and use metadata for real value
    private static let sexualActivityProtection: [String: Int] = [
        "protected": HKCategoryValue.notApplicable.rawValue,
        "unprotected": HKCategoryValue.notApplicable.rawValue,
        "unknown": HKCategoryValue.notApplicable.rawValue,
    ]

    /// Sleep session stage mappings
    /// Maps to HKCategoryValueIdentifier.sleepAnalysis values
    private static let sleepSession: [String: Int] = {
        // iOS 16.0+ — full map available
        if #available(iOS 16.0, *) {
            return [
                "inBed": HKCategoryValueSleepAnalysis.inBed.rawValue,
                "outOfBed": HKCategoryValueSleepAnalysis.asleepUnspecified.rawValue,  // Not directly mapped in iOS
                "sleeping": HKCategoryValueSleepAnalysis.asleepUnspecified.rawValue,  // Generic sleep state
                "awake": HKCategoryValueSleepAnalysis.awake.rawValue,
                "light": HKCategoryValueSleepAnalysis.asleepCore.rawValue,  // iOS uses asleep, we map light to asleep
                "deep": HKCategoryValueSleepAnalysis.asleepDeep.rawValue,
                "rem": HKCategoryValueSleepAnalysis.asleepREM.rawValue,
                "unknown": HKCategoryValueSleepAnalysis.asleepUnspecified.rawValue,
            ]
        }

        // Pre–iOS 16.0 — None of these symbols exist (it is guaranteed by the RecordTypeMapper)
        return [:]
    }()

    private static let vo2MaxMeasurementMethod: [String: Int] = [
        "cooperTest": HKVO2MaxTestType.maxExercise.rawValue,
        "rateRatio": HKVO2MaxTestType.predictionNonExercise.rawValue,
        "metabolicCart": HKVO2MaxTestType.maxExercise.rawValue,
        "multistageFitnessTest": HKVO2MaxTestType.maxExercise.rawValue,
        "fitnessTest": HKVO2MaxTestType.predictionSubMaxExercise.rawValue,
        "other": HKVO2MaxTestType.maxExercise.rawValue,
    ]

    // ------------------------------------------------------------------------
    // Registry of all category maps
    // ------------------------------------------------------------------------

    private static let registry: [String: [String: Int]] = [
        "CKBiologicalSex": biologicalSex,
        "CKBloodType": bloodType,
        "CKFitzpatrickSkinType": fitzpatrickSkinType,
        "CKIntermenstrualBleedingType": intermenstrualBleedingType,
        "CKContraceptiveValue": contraceptiveValue,
        "CKProgesteroneTestResult": progesteroneTestResult,
        "CKBodyTemperatureMeasurementLocation": bodyTemperatureMeasurementLocation,
        "CKMindfulnessSessionType": mindfulnessSessionType,
        "CKMenstruationFlow": menstruationFlow,
        "CKCervicalMucusAppearance": cervicalMucusAppearance,
        "CKOvulationTestResult": ovulationTestResult,
        "CKSexualActivityProtection": sexualActivityProtection,
        "CKVo2MaxMeasurementMethod": vo2MaxMeasurementMethod,
        // NOTE: SleepSession is not a category enum in dart (no `CK` prefix)
        "SleepSession": sleepSession,
    ]

    // ------------------------------------------------------------------------
    // Public API
    // ------------------------------------------------------------------------

    /// Decodes a Dart enum string into its corresponding iOS Int constant.
    ///
    /// - Parameters:
    ///   - category: The category name (e.g. "SkinTemperatureMeasurementLocation").
    ///   - value: The Dart enum value string (e.g. "wrist").
    /// - Returns: The matching iOS Int constant, or `nil` if unknown.
    public static func decode(categoryName: String, value: String) -> Int? {
        guard let categoryMap = registry[categoryName] else {
            return nil
        }
        return categoryMap[value]
    }

    /// Encodes an iOS Int constant into its corresponding Dart enum string.
    ///
    /// - Parameters:
    ///   - category: The category name (e.g. "SkinTemperatureMeasurementLocation").
    ///   - constant: The iOS constant value (e.g. 10 for wrist).
    /// - Returns: The Dart enum string, or `nil` if unknown.
    public static func encode(categoryName: String, constant: Int) -> String? {
        guard let categoryMap = registry[categoryName] else {
            return nil
        }
        // Reverse lookup — find key by value
        return categoryMap.first(where: { $0.value == constant })?.key
    }
}
